pylint checking documentation (#29906)
Azure/azure-sdk-for-python: doc/dev/pylint_checking.md

# Python SDK Pylint Guide

Cheat sheet for the Python SDK pylint general guidelines for your client library 

# Table of contents
  - [General Guidance](#general-guidance)
    - [What is Pylint?](#what-is-pylint)
    - [How to run Pylint?](#how-to-run-pylint)
  - [Ignoring Pylint Checkers](#ignoring-pylint-checkers)
  - [Pylint Warnings and Where to Find Them](#pylint-warnings-and-where-to-find-them)
    - [Next Pylint](#next-pylint)
  - [How to Prepare your SDK?](#how-to-prepare-your-sdk-for-a-new-pylint-update)

# General Guidance 

## What is pylint?

Pylint is a set of (mostly) [astroid](https://pylint.pycqa.org/projects/astroid/en/latest/index.html) based checkers that run static analysis on your code and check that your formatting aligns with Python's style guide (i.e [PEPs](https://peps.python.org/)). 

> In the Azure SDK for Python repository, in addition to the standard pylint library, there are also custom checkers, within the azure-pylint-guidelines-checker package, that help to customize our libraries to the standards described in the [Azure SDK for Python Guidelines](https://guidelinescollab.github.io/azure-sdk/python_introduction.html).

Author: ```suggestion
In the Azure SDK for Python repository, in addition to the standard pylint library, there are also custom checkers within the azure-pylint-guidelines-checker package that help to customize our libraries to the standards described in the [Azure SDK for Python Guidelines](https://guidelinescollab.github.io/azure-sdk/python_introduction.html).
```

## How to run Pylint?

One way to run pylint is to run at the package level with tox:

    .../azure-sdk-for-python/sdk/eventgrid/azure-eventgrid>tox -e pylint -c ../../../eng/tox/tox.ini

If you don't want to use tox, you can also install and run pylint on its own:

 - If taking this approach, in order to run with the pylintrc formatting and the custom pylint checkers you must also install the custom checkers and `SET` the pylintrc path. 

        pip install pylint
        pip install azure-pylint-guidelines-checker --index-url="https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-python/pypi/simple/"

        .../azure-sdk-for-python>SET PYLINTRC="./pylintrc"
        .../azure-sdk-for-python>pylint ./sdk/eventgrid/azure-eventgrid

>     Note that you may see different errors if running a different [version of pylint](https://github.com/Azure/azure-sdk-for-python/blob/fdf7c49ea760b1e1698ebbbac48794e8382d8de5/eng/tox/tox.ini#L90) than the one in CI.

Author: ```suggestion
    Note that you may see different errors if running a different [version of pylint or azure-pylint-guidelines-checker](https://github.com/Azure/azure-sdk-for-python/blob/fdf7c49ea760b1e1698ebbbac48794e8382d8de5/eng/tox/tox.ini#L90) than the one in CI.
```

  

# Ignoring Pylint Checkers

- If there is a pylint checker within your SDK that you wish to ignore for that specific scenario (i.e protected-access) you can ignore it with a comment on the offending line.

    `# pylint:disable=protected-access`  
    
- (Not Recommended) Another way to disable a checker is by ignoring the entire package in the [pyproject.toml](https://github.com/Azure/azure-sdk-for-python/blob/main/doc/eng_sys_checks.md#the-pyprojecttoml).

# Pylint Warnings and Where to Find Them 

Information on the custom pylint checkers resides [here](https://github.com/Azure/azure-sdk-tools/blob/main/tools/pylint-extensions/azure-pylint-guidelines-checker/README.md). [This table](https://github.com/Azure/azure-sdk-tools/blob/main/tools/pylint-extensions/azure-pylint-guidelines-checker/README.md#rules-list), provides information on each custom check and how to resolve them. 

In addition to being a part of the CI, the custom pylint checkers are also integrated into ApiView. If there are unaddressed warnings, they will show as system comments. 

## Next Pylint

> There is now a new step on the CI pipeline called `Run Pylint Next`. This is merely a duplicate of the `Run Pylint` step with the exception that `Run Pylint Next` uses a newer version of pylint and a newer version of the custom pylint checkers.

Reviewer: does newer version == latest?

Author: yup let me change that!

This next-pylint enviornment can also be run locally through tox:

    tox -e next-pylint -c ../../../eng/tox/tox.ini

> The errors generated by the `Run Pylint Next` step will not break your weekly test pipelines, but still you should take note and fix them when you have the time. 

Author: ```suggestion
The errors generated by the `Run Pylint Next` step will not break your weekly test pipelines, but you should still fix them when you have the time so that your client library is up to date for the next release. 
```

# How to prepare your SDK for a new pylint update?

> Check each client libraries `Run Pylint Next` output in the [test-weekly CI pipeline](https://dev.azure.com/azure-sdk/internal/_build?pipelineNameFilter=python%20*%20tests-weekly). If there is no corresponding test-weekly pipeline, run `next-pylint` locally with `tox` as described in [How to run Pylint?](#how-to-run-pylint). In order to ensure that the SDK pipeline will not break when pylint is updated, make sure to address all pylint warnings present. 

Author: ```suggestion
Check that each client library's `Run Pylint Next` output in the [test-weekly CI pipeline](https://dev.azure.com/azure-sdk/internal/_build?pipelineNameFilter=python%20*%20tests-weekly). If there is no corresponding test-weekly pipeline, run `next-pylint` locally with `tox` as described in [How to run Pylint?](#how-to-run-pylint). In order to ensure that the SDK pipeline will not break when pylint is updated, make sure to address all pylint warnings present. 
```

# Python SDK Pylint Guide

> Cheat sheet for pylint general guidelines in the Python client SDK library. 

Reviewer 2: Maybe something like this ?

```suggestion
Cheat sheet for the Python SDK pylint general guidelines for your client library 
```

# Table of contents
  - [How to Prepare your SDK?](#how-to-prepare-your-sdk-for-a-new-pylint-update)
  - [General Guidance](#general-guidance)
    - [What is Pylint?](#what-is-pylint)
    - [How to run Pylint?](#how-to-run-pylint)
  - [Ignoring Pylint Checkers](#ignoring-pylint-checkers)
  - [Pylint Warnings and Where to Find Them](#pylint-warnings-and-where-to-find-them)
    - [Next Pylint](#next-pylint)

# How to prepare your SDK for a new pylint update?

> Check each SDK's `Run Pylint Next` output in the [test-weekly CI pipeline](https://dev.azure.com/azure-sdk/internal/_build?pipelineNameFilter=python%20*%20tests-weekly). If there is no corresponding test-weekly pipeline, run `next-pylint` locally with `tox` as described in [How to run Pylint?](#how-to-run-pylint). In order to ensure that the SDK pipeline will not break when pylint is updated, make sure to address all pylint warnings present. 

Reviewer 2: ```suggestion
Check each client libraries `Run Pylint Next` output in the [test-weekly CI pipeline](https://dev.azure.com/azure-sdk/internal/_build?pipelineNameFilter=python%20*%20tests-weekly). If there is no corresponding test-weekly pipeline, run `next-pylint` locally with `tox` as described in [How to run Pylint?](#how-to-run-pylint). In order to ensure that the SDK pipeline will not break when pylint is updated, make sure to address all pylint warnings present. 
```

See below for more information on pylint checkers and how to ignore warnings if needed.

# General Guidance 

### What is pylint?

Pylint is a set of (mostly) [astroid](https://pylint.pycqa.org/projects/astroid/en/latest/index.html) based checkers that run static analysis on your code and check that your formatting aligns with Python's style guide (i.e [PEPs](https://peps.python.org/)). 

In the Azure SDK for Python repository, in addition to the standard pylint library, we added custom checkers that help to align our libraries with the [Azure SDK for Python Guidelines](https://guidelinescollab.github.io/azure-sdk/python_introduction.html).

### How to run Pylint?

One way to run pylint is to run at the package level with tox:

    .../azure-sdk-for-python/sdk/eventgrid/azure-eventgrid>tox -e pylint -c ../../../eng/tox/tox.ini

> If you don't want to use tox you can also install and run pylint on its own:

Reviewer 3: baby nit:
```suggestion
If you don't want to use tox, you can also install and run pylint on its own:
```

 - If taking this approach, in order to run with the pylintrc formatting and the custom pylint checkers you must also install the custom checkers and `SET` the pylintrc path. 

- More info on how to run the custom pylint checkers [here](https://github.com/Azure/azure-sdk-tools/blob/main/tools/pylint-extensions/pylint-guidelines-checker/README.md)

>         pip install pylint

Reviewer 3: did you mean to have the commands here in addition to the link? if not, maybe you can just pick one. or if you did mean to add a link for additional information, maybe you can move the link to a link under the commands and say something like `Additional information about custom checkers [here]`.

        pip install pylint-guidelines-checker --index-url="https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-python/pypi/simple/"

        .../azure-sdk-for-python>SET PYLINTRC="./pylintrc"
        .../azure-sdk-for-python>pylint ./sdk/eventgrid/azure-eventgrid

    Note that you may see different errors if running a different version of pylint than the one in CI.

# Ignoring Pylint Checkers

There are three ways to ignore pylint checkers:

> - If there is a pylint checker within your SDK that you do not think applies, or that you wish to ignore for that specific scenario (i.e protected-access) you can locally ignore it on the specific line of the warning with the comment `# pylint:disable=protected-access`. If a pylint warning appears multiple times within a file you can ignore all related warnings in that file by putting the disable comment at the top of the file.

Reviewer 3: This is a lot of text to read through for documentation. I wonder if it might be nice to have the commands/examples listed on separate lines, b/c that catches the eye immediately. Like what the typing doc has here: https://github.com/Azure/azure-sdk-for-python/blob/main/doc/dev/static_type_checking.md#how-to-ignore-type-checking-errors

> - Another way to disable a pylint checker is by putting in under the `disable` list in the `pylintrc` file. This should only happen if the checker is giving too many false positives, or is not applicable to our general library. Any additions to the pylintrc file should be discussed beforehand. 

Reviewer 3: a link to an example pylintrc file would be helpful

> - Another way to disable a pylint checker is by putting in under the `disable` list in the `pylintrc` file. This should only happen if the checker is giving too many false positives, or is not applicable to our general library. Any additions to the pylintrc file should be discussed beforehand. 

Reviewer: Suggest we don't document this as an option. If we do put something under the disable list, it will probably be discussed internally within our team / added by our team.

> - (Not Recommended) The last way to disable a checker is by ignoring the entire package in the [environment_exclusions.py](https://github.com/Azure/azure-sdk-for-python/blob/main/tools/azure-sdk-tools/ci_tools/environment_exclusions.py) or as a temporary fix ignoring parts of the package via the pylintrc `ignore-paths` option. 

Reviewer 3: link to an `ignore-paths` example would be helpful

> - (Not Recommended) The last way to disable a checker is by ignoring the entire package in the [environment_exclusions.py](https://github.com/Azure/azure-sdk-for-python/blob/main/tools/azure-sdk-tools/ci_tools/environment_exclusions.py) or as a temporary fix ignoring parts of the package via the pylintrc `ignore-paths` option. 

Reviewer: The pyproject.toml has superseded the environment exclusions list. Might want to link to this: https://github.com/Azure/azure-sdk-for-python/blob/main/doc/eng_sys_checks.md#the-pyprojecttoml

        

# Pylint Warnings and Where to Find Them 

Info on the custom pylint checkers [here](https://github.com/Azure/azure-sdk-tools/tree/main/tools/pylint-extensions/azure-pylint-guidelines-checker).

> In addition to being a part of the CI, the custom pylint checkers are also integrated into ApiView and will show as warnings there. 

Reviewer 3: if possible, might be nice to link to an example of what a warning in apiview would look like or maybe add a screenshot here

## Next Pylint

> There is now a new step on the CI pipeline called `Run Pylint Next`. This is merely a duplicate of the `Run Pylint` step with the exception that `Run Pylint Next` has the newer version of pylint and the newer version of the custom pylint checkers. 

Reviewer 2: ```suggestion
There is now a new step on the CI pipeline called `Run Pylint Next`. This is merely a duplicate of the `Run Pylint` step with the exception that `Run Pylint Next` has the newer versions of pylint and the custom pylint checkers. 
```
